endpoint: don't hold the endpoint lock while generating policy

squeed opened the pull request:
This PR has, basically, two parts. For more details, see the individual commit messages.

1. Make some endpoint accessor methods lock-free. Some are replaced with  sync.Atomic, others are actually immutable and don't need locking.
2. Calculate an endpoint's policy without holding the endpoint lock.

As preparation for further refactors of the policy engine, it would be desirable to stop holding the whole endpoint lock while we compute policy. To do this, we need to read some values, unlock, compute policy, then lock and apply. More subtly, we need to make sure that regeneration succeeds before unlocking again.

## Why is this safe?

Good question.

Endpoint regeneration has a complicated locking story. There are two interesting locks here, both of which combine to protect certain fields that are modified both internally and externally.

Regeneration is serialized by the `ep.buildMutex`, which ensures we never regenerate and compute policy in parallel. That means we are free to read and write the internal policy fields freely, since we never have to protect against write collisions.

Other fields accessed externally may be protected behind `ep.mutex`. However, this is too heavy-weight. Why? Because during policy computation, the policy engine needs to read back in to the endpoint. Additionally, the ipcache needs to provide policy deltas to the existing policy without being blocked.

The solution is to determine, exactly, which fields need to be blocked by which mutex. The critical observation is that **some fields need both the mutex and the buildMutex to be written to**. This ensures consistency in the face of regeneration and parallel operations.

Additionally, the critical observation is that during the policy calculation process, the L4Policies themselves register for incremental policy updates. That means that we have no chance of missing an incremental update during policy calculation, and it is safe to let the ipcache make progress. Specifically, the order is:

1. Create CachedSelectors. These subscribe for incremental updates
2. Call UpdatePolicy() -> resolvePolicylocked(), which creates the `L4Filters` and allows them to start reading incremental updates
3. Call Consume() -> DistillPolicy(), which evaluates the cached selectors for a list of applicable identities.
4. Lock the endpoint, which blocks ipcache progress
5. Apply the policy, then call ApplyPolicyMapChanges, which consumes any incremental updates that accumulated between steps 3 and 4.

Since incremental updates are "idempotent", it is safe to see that we won't miss any incremental updates, even if the ipcache makes progress during the policy computation / endpoint regeneration process.

### Relevant fields and their mutexes:

- `ep.desiredPolicy` -- READ by the ipcache, WRITTEN by regeneration. Protected by `ep.mutex`
- `ep.policyRevision` -- READ by everyone, WRITTEN by regeneration AND the policy engine. Protected by `ep.mutex` and `ep.buildMutex`
- `ep.selectorPolicy` -- READ by regeneration, WRITTEN by regeneration *and* the identity resolution loop. Protected by `ep.mutex` and `ep.identityRevision`
- `ep.realizedRedirects` -- READ external policy calculation, WRITTEN by regeneration. Protected by `ep.mutex` and `ep.buildMutex`

### Processes running in parallel

- Endpoint regeneration. Takes `ep.buildMutex`, does policy calculations, then takes `ep.mutex`. Commits and drops all locks at the end.
- ipcache. Needs to call `ep.ApplyPolicyMapChanges`, which requires `ep.mutex`
- identity resolution. Clears `ep.selectorPolicy`, holds `ep.mutex`
- policy ingestion. When a policy is a no-op for an endpoint, calls `ep.SetPolicyRevision()` which requires `ep.mutex` and `ep.buildMutex`.

squeed committed:
endpoint: don't hold the endpoint lock while generating policy

As preparation for other refactors of the policy engine, no longer hold
the endpoint lock while calculating policy. This is safe to do, since
the only input is the endpoint's security identity. Furthermore, if,
somehow, policy were to be calculated in parallel, we can reject an
update if its revision is too old.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
pkg/endpoint: make GetNamedPorts lock-free

This function is called deep within the policy generation hierarchy, and
is at a risk of causing deadlocks.

Given that it's just reading a pointer to a never-mutated map, we can
safely stash this behind an atomic Pointer and remove the lock.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
pkg/endpoint: make some more accessor methods lock-free

It turns out that most of the endpoint identities, e.g. pod name /
namespace, are actually immutable. So, there's no need to grab a lock
before reading them.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
pkg/proxy: mechanical: remove unused methods from interface

These methods are no longer used; remove them from the
EndpointInfoSource interface.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

tommyp1ckles commented on pkg/endpoint/endpoint.go:
```diff
@@ -301,8 +302,7 @@ type Endpoint struct {
 	// proxy port number.
-	// You must hold Endpoint.mutex to read or write it.
-	realizedRedirects map[string]uint16
+	realizedRedirects sync.Map
```
Changing this code will be a tiny bit less readable/safe without the typed map, I wonder if we should have a thin typed generic wrapper around this type.

squeed replied:
That would certainly be useful in this case. I wonder if there's one in the standard-ish library (/x, /pkg) already.

squeed replied:
I've decided to revert this part of the change.

tommyp1ckles commented on pkg/endpoint/endpoint.go:
```diff
@@ -301,8 +302,7 @@ type Endpoint struct {
 	// proxy port number.
-	// You must hold Endpoint.mutex to read or write it.
-	realizedRedirects map[string]uint16
+	realizedRedirects sync.Map
```
I don't fully understand the semantics of sync.Map, I know it's only advantageous in certain access patterns.  Could we add a comment explaining the specific case of endpoints and why this an advantage in this case?

squeed replied:
You know, it could probably be an atomic.Pointer instead, which would preserve type and make things generally easier.

squeed replied:
It's good you asked this question.

As part of writing out the access patterns, I realized we could do this in a much cleaner way, just with some extra comments around locking. So, this PR will change a bit. Thanks for that!

tommyp1ckles commented on pkg/endpoint/endpoint.go:
```diff
@@ -1210,7 +1207,6 @@ func (e *Endpoint) SetK8sNamespace(name string) {
 // so that the map can be used concurrently without keeping locks.
-// Reading the 'e.k8sPorts' member (the "map pointer") *itself* requires the endpoint lock!
 // Can't really error out as that might break backwards compatibility.
 func (e *Endpoint) SetK8sMetadata(containerPorts []slim_corev1.ContainerPort) error {
```
nit: we can probably remove the error return value on this, I don't think it ever actually returns non-nil.

squeed replied:
oh nice, for sure.

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -769,6 +777,12 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 	}
 
+	// Set the computed policy as the "incoming" policy
+	// this can fail if the policy revision was bumped during calculation
```
```
// Set the newly computed policy as the "incoming" policy.
// A new policy revision could preempt this new policy (`policyResult`), causing the following function to return an error. This is OK because endpoint regeneration will run again.
```

Just some clarifying language to be more explicit.  The last sentence I added I'm not 100% sure about. Is that true? It would be nice to tell the code reader what happens when a new policy revision comes in and `setDesiredPolicy` returns an error.

squeed replied:
My comments were wrong wrong wrong anyways. I've written a bit about locking in the PR description, and I'm going to push some of that in to code comments now.

christarazi commented on pkg/endpoint/policy.go:
```diff
@@ -143,79 +150,143 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+// The endpoint write lock must be held, and not released until the desired policy has
+// been pushed in to the policymaps. This is so that we block ApplyPolicyMapChanges, which
+// has the effect of blocking the ipcache from updating the ipcache bpf map. It is required
+// that any accumulated changes are pushed in to the policymap before the ipcache map.
```
:ok_hand: 

To expand, I think it would help to mention in the comment what the consequences are if someone were to break that constraint: transient policy drops.

squeed replied:
done.

christarazi commented on pkg/endpoint/policy.go:
```diff
@@ -143,79 +150,143 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+			return nil, err
 		}
 	}
-	// TODO: GH-7515: This should be triggered closer to policy change
-	// handlers, but for now let's just update it here.
-	if err := repo.GetPolicyCache().UpdatePolicy(e.SecurityIdentity); err != nil {
```
This todo is not actually resolved. The todo is basically saying that policy computation can be done outside of the endpoint completely. While this overall PR gets us closer in that direction, we are technically still in the context of the endpoint. 

Going forward, we can resolve #7515 when we can calculate policy for endpoints (outside of the context of a single endpoint) when
1. policies change
2. identities change
3. endpoints change

and "change" here means added or removed. 

Concretely what this could look like is some `func CalculatePolicy(identity *identity.Identity)` which we call on each of the above listed occurrences. Then the only logic necessary in the endpoint should just be a lookup like 

```
selectorPolicy := repo.PolicyCache().Lookup(e.SecurityIdentity)
...
if forcePolicyCompute {
  repo.PolicyCache().UpdatePolicy(e.SecurityIdentity)
}
```

I toyed with this in https://github.com/squeed/cilium/compare/policy-allocate-when-used-v2...christarazi:pr/christarazi/separate-policy-calc-ep-regen, but obviously needs more work. Just wanted to share more context.

squeed replied:
aha, I saw that the issue was closed and didn't really go much further than that. If we wanna do it, we shouldn't'a closed the issue :-).

Not to get too distracted from the current issue, but I'm not sure that just extracting policy computation is that interesting. More useful would be the ability to fully push policy to bpf & Envoy outside of a regeneration. I don't know how interleaved they are, but it seems like it could be doable.

christarazi replied:
> Not to get too distracted from the current issue, but I'm not sure that just extracting policy computation is that interesting. More useful would be the ability to fully push policy to bpf & Envoy outside of a regeneration. I don't know how interleaved they are, but it seems like it could be doable.

Agreed, and I think we need to extract / decouple policy calculation from endpoint regeneration in order to be able to push policy to BPF & Envoy outside of a regeneration, in order to simplify the model. Today, the logic is a bit interleaved and harder to reason about. 

A regeneration is considered a full policy calculation, while `ApplyPolicyMapChanges` is incremental. If we separate policy calculation completely from endpoint, then that should in theory allow us to _always_ perform incremental policy updates, which should improve performance, overhead, and reduce potential interleaving of critical sections. It will also put the code into a structure where it's easier to introduce `pkg/stream` infrastructure.

The way I see it is we can have a model where `CalculatePolicy(identity)` returns a policy revision and the desired computed policy, and hands that to each endpoint that has `identity`. Each endpoint with that `identity` will then take their realized, current policy and compute the diff between the desired policy from `CalculatePolicy`, and implements the diff. The endpoints would then be considered implementing the returned policy revision.

In contrast to today, each endpoint attempts to compute the policy diff. `N` endpoints -> compute policies `N` times (there's some caching of course). 

We sort of "invert" this relationship with `CalculatePolicy` by doing: Policy computed once -> `N` endpoints updated. 

This makes the model much simpler to think about and should enable us to always have incremental policy updates.

squeed replied:
All of this is a fine idea - and would indeed be cleaner - except for the specter of named ports. This is basically what forces policy calculation to be done in two phases, an identity-generic phase and and an endpoint-specific phase.

Ultimately, we don't currently detect the "edge" that is an Identity being assigned to an Endpoint on the node. If we build such a thing in to the EndpointManager, then that seems like the right place to move some of these pure control-plane computations.

jrajahalme replied:
This reminds me of an aspect of named ports being resolved per-endpoint after a shared `selectorPolicy` has been computed: It is possible for a named port resolve to the same port that is used in the policy by number, and in the extreme case it would be possible for both of them specify L7 policies for different proxy parsers. Which one gets plumbed to the datapath/Envoy is a matter of map iteration order. Otherwise I think we handle this gracefully, e.g.:
- the policy with a deny rule gets to the datapath, if both are denies that is not a problem
- if one is redirect while other is not, the one with redirect takes precedence

christarazi replied:
> All of this is a fine idea - and would indeed be cleaner - except for the specter of named ports. This is basically what forces policy calculation to be done in two phases, an identity-generic phase and and an endpoint-specific phase.

Agreed, good point. It would be really nice to make that two-phase policy computation (identity-based & named port-based) very clear in the code. I think it's very doable.  

One way to start would be to identify which named port-specific code exists in the policy computation and see how we can extract it out. 

I'm also thinking of another possibility where we could potentially even compute the named port policy within the identity-based computation. Consider that identities can map to multiple endpoints, one of those endpoints might have a named port policy. So you could imagine that the identity-based policy can detect this case, generate the named port policy as well, and then provide it to the endpoint that needs it. When endpoints apply the policy that was generated for them, they can perform some lookup to detect whether they need named ports policy, and then apply it as well if so. 

> Ultimately, we don't currently detect the "edge" that is an Identity being assigned to an Endpoint on the node. If we build such a thing in to the EndpointManager, then that seems like the right place to move some of these pure control-plane computations.

Yep, we can definitely take steps to improve that. :+1:

christarazi commented on pkg/policy/distillery.go:
```diff
@@ -139,6 +139,8 @@ func (cache *PolicyCache) LocalEndpointIdentityRemoved(identity *identityPkg.Ide
 // Lookup attempts to locate the SelectorPolicy corresponding to the specified
 // identity. If policy is not cached for the identity, it returns nil.
+//
+//nolint:staticcheck // Because this causes a spurious error in endpoint/policy.go
```
Kinda unfortunate that we need to apply nolint here as well. Does this effectively disable all linting in this function?

squeed replied:
It's super frustrating, especially because the "lint error" at this line is only emitted as an informational bit to explain a (spurious) issue at a caller. But, the tooling isn't smart enough to ignore the informational hint and fails lint without the tag here as well.

This disables the staticcheck linter for this function. Given that it's a one-line trivial function, it's not a big deal. Once the heat is off, I'm going to try and come up with a minimal reproducer and file a bug.

squeed replied:
update: the lint error doesn't show up if the lval in question is a structure field (maybe because it can't determine shared access?) so these tags can be removed, but not for a good reason.

christarazi reviewed (commented):
Nice work! A couple of comments.

jrajahalme commented on pkg/endpoint/endpoint.go:
```diff
@@ -1173,35 +1178,27 @@ func (e *Endpoint) leaveLocked(proxyWaitGroup *completion.WaitGroup, conf Delete
-	e.unconditionalLock()
+// SetK8sNamespaceLocked modifies the endpoint's pod name
+// only used by testing code.
+func (e *Endpoint) SetK8sNamespaceLocked(name string) {
```
It would be better to remove non-production code, e.g, by moving it to the tests that need it, or refactoring the tests to remove the need for the (non-essential?) function.

squeed replied:
done.

jrajahalme commented on pkg/endpoint/identifiers.go:
```diff
@@ -17,25 +17,22 @@ func (e *Endpoint) GetContainerName() string {
-	e.unconditionalLock()
+// SetContainerNameLocked modifies the endpoint's container name
+// Only used during tests.
+func (e *Endpoint) SetContainerNameLocked(name string) {
```
Would be better move code not used in production to tests, if possible. Maybe these kinds of fields could be classified as "immutable after creation" and then make calculated violations of that policy in tests if/when needed?

jrajahalme commented on pkg/endpoint/identifiers.go:
```diff
@@ -17,25 +17,22 @@ func (e *Endpoint) GetContainerName() string {
 // Kubernetes pod
 func (e *Endpoint) GetK8sPodName() string {
-	e.unconditionalRLock()
 	k8sPodName := e.K8sPodName
```
I'd like to see a comment on this line to the effect `// const after creation` so that the code would be easier to read without needing to refer back to the field definition to check if lock needs to be taken each time reading this code?

squeed replied:
fair enough, can do for all the relevant fields.

jrajahalme commented on pkg/endpoint/identifiers.go:
```diff
@@ -46,34 +43,29 @@ func (e *Endpoint) HumanStringLocked() string {
-	e.unconditionalRLock()
-	defer e.runlock()
-
 	return e.getK8sNamespaceAndPodName()
```
Do we still need both exported and un-exported function of the "same" name and function?

squeed replied:
hah. No.

jrajahalme commented on pkg/endpoint/identifiers.go:
```diff
@@ -46,34 +43,29 @@ func (e *Endpoint) HumanStringLocked() string {
-	e.unconditionalLock()
+// SetK8sPodNameLocked modifies the endpoint's pod name
+// Only used for tests.
+func (e *Endpoint) SetK8sPodNameLocked(name string) {
```
Would be better to move non-production code to tests, or better yet eliminate the need for this in the tests by using production code for creating the endpoints also in tests, if possible?

squeed replied:
I agree completely; unfortunately the tests that reference this are out-of-package. I was trying to make this PR as minimal as possible.

The tests that use this functionality could _probably_ be refactored to stop, I can check.

squeed replied:
There, I managed to remove some of the silly setter methods.

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -735,8 +735,18 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
 
+	// regenerate policy without holding the lock.
+	// This is because policy generation needs the ipcache to make progress, and the ipcache needs to call
+	// endpoint.ApplyPolicyMapChanges()
```
Care should be taken with applying incremental policy updates while calculating the endpoint policy, as they are functionally interdependent via code like this:
```
			if e.desiredPolicy == e.realizedPolicy {
				// Any map updates when a new policy has not been calculated are taken care by incremental map updates.
				continue
			}
```

In the worst case an incremental change may get "lost" due to (full) policy update not seeing it yet, but then the incremental change not being pushed to the datapath due to parallel executing full policy update apparently having done a "full" update.

jrajahalme replied:
I checked the code again, did not recall that we pull in incremental changes that may have happened during the policy computation right before we sync the bpf policy map.

squeed replied:
Yeah, this stuff is subtle. I tried to clarify this in the comments around locking, but there's always more to be written.

jrajahalme commented on pkg/endpoint/policy.go:
```diff
@@ -141,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+	securityIdentity := e.SecurityIdentity
+
+	result := &policyGenerateResult{
+		selectorPolicy:   e.selectorPolicy,
```
Selector policy would better be handled separately, having it here gives the impression it is something that can change on every policy computation, while it is only ever changed if it is still `nil`. Also, it should remain the same value for a given security identity, even if looked up again.

If possible, leave it out from `policyGenerationResult` and set it earlier in the code while still holding the necessary locks and bail out if it is still nil before any policy computation is done?

squeed replied:
Right, so, selectorPolicy is written in two places - after the first policy computation, where is set, or nilled out when the SecurityIdentity is changed.

Given that identity can change during policy calculation, we shouldn't write to selectorPolicy until we're sure that identity hasn't been changed out from under us. I agree this is a bit awkward, but I can't think of an easier way to do this.

jrajahalme replied:
Right, I had to find the spot where it is nilled to fully grasp this.

jrajahalme reviewed (changes requested):
First 3 commits look safe enough, but the final one feels too risky to push into a release this late in the release cycle. Please factor out any bug fixes to a separate PR, and let's hold on the major change until the feature freeze for 1.14 is over.

jrajahalme reviewed (approved):
Went through the DNS proxy logic pushing FQDN selector updates and then triggering endpoint regenerations to process incremental policy updates, did not find any issues with it w.r.t this change.
One more general thing to note is that `Endpoint.ApplyPolicyMapChanges()` calls `updateNetworkPolicy` after applying the map changes for historical reasons, I think. We used to deduce the need for network policy updates from the contents of the incremental map updates, but were then hit by 3 code paths competing for consuming the same updates (policy regen, map sync controller, and DNS proxy code paths all can call endpoints `applyPolicyMapChanges`), so any one of them is not guaranteed to see the whole set of related map updates and it was impossible to deduce the need for (Envoy) network policy updates from the map changes applied at any one spot. But we still update the network policies here after map updates, while the policy everywhere else is to update network policies first, map updates second.

squeed committed:
test/MockIdentityAllocator: Sanitize the generated ID

This ensures the generated ID works like IDs allocated normally - that
their LabelArray is set.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
endpoint: add policy engine race test

This adds a small test that ensures incremental updates are never lost,
even in the face of significant identity churn.

It simulates a churning ipcache flinging identities in to the policy
engine, and similarly recalculates policy constantly.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

christarazi commented on pkg/endpoint/policy_test.go:
```diff
@@ -38,3 +53,163 @@ func (s *EndpointSuite) TestUpdateVisibilityPolicy(c *check.C) {
+			haveIDs.Insert(identity.NumericIdentity(k.Identity))
+		}
+
+		// It is okay if we have *more* IDs than allocatedIDs, since we may have propagated
+		// an ID change through the policy system but not yet added to the extra list we're
```
> since we may have propagated an ID change through the policy system

Specifically where could that have happened? Mostly just curious so we can explain that in the comment as well because it's not obvious to me.

squeed replied:
In order to simulate a "churning" ipcache, the test only stops the identity allocation loop to write to the `allocatedIDs` set. This means that, depending on the interleaving, we may have fully plumbed-through one more ID than is written to the set. In other words `len(have) == len(want) + 1`. I could add more locking to prevent this, but then the test wouldn't be as faithful to the actual system.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
+//   - ep.policyRevision: ep.mutex and ep.buildMutex must be held to write to this, or else
+//     the policyRevision may go backwards from no-op bumps from policy ingestion
```
How is this possible?

Policy revision bumps should be triggered in `reactToRuleUpdates()` -> `epsToBumPRevision.ForEachGo(...)` -> `epp.PolicyRevisionBumpEvent()` -> Endpoint eventqueue -> set the policy revision. The eventqueue is the same eventqueue that handles full policy regenerations for when newly ingested policies _do_ affect the endpoint. So it should be linear, the endpoint policy revision bump should never occur for a lower revision than a revision already calculated during an actual policy calculation event :thinking: 

(Incidentally, the reason I dug into this is that this implies that all policy ingestion into the policy repo now relies on all endpoints catching up with policy calculation. ~I didn't see that consequence getting called out anywhere, but it feels like something that should be acknowledged and discussed~ Correction, I see it now in the PR description but I didn't catch it in the code or commit message)

squeed replied:
Hmm, interesting. I'll do some more digging. I *definitely* saw consistent test failures around endpoints failing to reach a certain policy revision, and the solution was *more locking*, which 100% eliminated that failure.

squeed replied:
Well, I can't reproduce the failure now, even after removing the lock. I'll update the inaccurate comment. There must have been some other incorrect logic also cleaned up.

squeed replied:
@joestringer this brings up another question: how much can / should we rely on exported methods only being called by the event queue?

joestringer replied:
Yeah good point, ideally the methods would be private so there's no other path. But at the same time, I think that if we don't reinforce that assumption, then we open a whole can of worms around the style of the Endpoint internals because we have to know how the entire rest of the daemon may end up calling into Endpoint. This thread is one such example, there are surprising interactions between the way policy ingestion and internal endpoint policy calculation works.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+// updateRealizedState have been called
+//
+// There are a few fields that depend on this exact configuration of locking:
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
```
Incidentally do you have a particular interaction in mind that may trigger policy drops in such a scenario? Is this another ipcache incremental policy update logic synchronization issue?

squeed replied:
I don't think this is an issue, just a potential inconsistency that `ep.mutex` has always protected, and I was making explicit.

If we didn't have `ep.mutex`, we could get in the situation where `ep.desiredPolicy` has been updated, but we've not yet called `ep.syncPolicyMap()`. When the ipcache calls `ep.ApplyPolicyMapChanges`, it consumes all deltas queued in `ep.desiredPolicy`. _If we didn't have the endpoint lock_, something like this could happen.

0. `ep.realizedPolicy` selects Identity 1
1. Identity 1 is replaced by 2. The ipcache calls `UpdateIdentities(added=2, deleted=[])`, 
2. An endpoint recomputes policy. The policy selects identities 1 and 2.
3. The new policy is written to `ep.desiredPolicy` **but not synced yet**
4. The ipcache calls `ep.ApplyPolicyMapChanges`, which is a no-op because `ep.desiredPolicy` already has identity 2. At this point, the BPF PolicyMap only allows ID 1.
5. The ipcache swaps the Identity for IPs from 1 to 2. At this point, there could be a policy drop, since 2 is not in the endpoint's BPF PolicyMap.
6. The endpoint calls `ep.syncPolicyMap()` which corrects the drop.

At this point we have a policy drop, since we never added identity 2 to the BPF map. Again, **this was never a bug**, since we lock before step 3, ensuring that steps 3 and 6 must happen before steps 4 and 5. I just documented this requirement, since there were concerns about parallel ipcache operations and missed updates.

joestringer replied:
I wish I had a suggestion on how we might encode this into the commit message or commit content since it seems like a super helpful example, but I'm not quite sure the best format to cover this, other than just having the comment we already have here. :sweat_smile: 

I guess we might consider adding something like ".... policy drops may occur due to the lack of synchronization between full policy generation and incremental policy generation".

squeed replied:
I added some more words, hopefully those are useful to whomever next stumbles down this path.

joestringer replied:
Did those land in the current version? The changes are not entirely obvious to me. Though I'm also fine with this in general :+1:

joestringer replied:
Ah I see they ended up at `setDesiredPolicy()` :+1:

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+//     the policyRevision may go backwards from no-op bumps from policy ingestion
+//   - ep.selectorPolicy: this may be nulled if the endpoints identity changes; we must
+//     check for this when committing. ep.mutex must be held
+//   - ep.realizedRedirects: this is read by external callers as part of policy generation,
+//     so ep.mutex must not be required to read this. Instead, both ep.mutex and ep.buildMutex
```
I looked this up, because I was surprised. I agree that I see `DistillPolicy()` calling into this. This surprises me because distilling the policy is part of policy calculation, ie figuring out the _desired_ state. It seems odd to me for desired state calculation to take into account the realized state.

The only thing I could reason about here is that maybe the initial realization of the redirects is done through some other path, then the desired policy calculation takes this into account in order to reuse the cached port / avoid reallocating the port. Still seems somehow wrong to me, there should just be a port management facility that is able to provide that guarantee so that the policy engine doesn't have to reach into realized state to pull out the desired state.

squeed replied:
I believe this is used to preserve re-usable redirects across policy revisions. There also is a stateful redirect manager, I'm not sure why the endpoint also has to provide that state. @jrajahalme?

(The interesting bits are in `ep.addNewRedirectsFromDesiredPolicy()` which calls `proxy.CreateOrUpdateRedirect()`

joestringer replied:
(to be clear, I don't expect you to address this in this PR, just raising the discussion point since it surprised me)

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -735,8 +735,18 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
+	// regenerate policy without holding the lock.
+	// This is because policy generation needs the ipcache to make progress, and the ipcache needs to call
+	// endpoint.ApplyPolicyMapChanges()
+	stats.policyCalculation.Start()
+	policyResult, err := e.regeneratePolicy()
+	stats.policyCalculation.End(err == nil)
```
I realize you're probably just propagating this forward from the previous implementation, but inside `e.regeneratePolicy()` it's already accounting `stats.policyCalculation`. It shouldn't be in both places, probably we can remove it from the caller here.

squeed replied:
Hmm, they're separate statistics - one is a `policyRegenerationStatistics{}` and the other is `regenerationStatistics{}` :-)

joestringer replied:
Oh god.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+	}
+
+	// Copy out some values we care about, then unlock
+	forcePolicyCompute := e.forcePolicyCompute
```
Minor tricky part here: Previously, we held the lock, chose whether to force policy computation, then stored false (in `setDesiredPolicy()`), then released. However, I think that now that we fetch this outside the lock, it's possible for us to calculate policy for an old identity, then identity labels change (which would also set `e.forcePolicyCompute`), then call `setDesiredPolicy` with both (a) the old identity's policy and (b) determining that the next policy generation doesn't need to force policy computation. But actually it _does_ still need to force policy computation.

It feels like we almost need to clear the flag before calculating policy, then calculate policy, then iff (policy calculation failed && forcePolicyCompute was set), grab the lock and set the flag again. Plus remove the flag clearing in `setDesiredPolicy()`. Basically this way once we release the lock in `regeneratePolicy()` it's possible for some other caller to come in and ensure that the _next_ regeneration will also force policy calculation, taking into account any changes that may have occurred while policy calculation was not holding the lock.

squeed replied:
We already catch the first case (identity reset) by checking to see if `ep.identityRevision` changed during computation. So, we should be good in that regard.

I think you're right; we should clear forcePolicyCompute in the first phase. Good idea, I'll do that.

joestringer replied:
Ah yes I did see the identityRevision, didn't account for that in this line of reasoning. :+1:

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
 		}).Debug("Skipping unnecessary endpoint policy recalculation")
-
-		return nil
+		repo.Mutex.RUnlock()
```
nit: I can't help but wonder if it'd be worth restructuring some of the function here a little in order to make use of defers for the repo mutex unlock. The code is correct, but sometimes developers can add a new error condition and neglect to properly handle locking when the lock release is spread throughout the function. Nothing a good round of review can't iron out, but from memory we have previously shipped error paths that don't unlock. It's usually not pretty, even if relatively easily found with Go tooling ;-) .

squeed replied:
If we can't leave a few landmines for future developers, what's even the point?

squeed replied:
In all seriousness, I'll think about it for a bit. We did previously hold on to `repo.Mutex` for way too long

squeed replied:
Besides adding yet another function, I don't see an obvious fix. It would be good to drop `repo.Mutex` before `policy.Consume()` as, distillation is expensive.

joestringer replied:
Yeah no worries. Again, this one's a nit, so up to your judgement. I'm fine with this as is, we can always iterate further on the code in subsequent PRs.

joestringer reviewed (changes requested):
Couple of minor issues with stats and forcePolicyCompute flag. I also raised a few points for discussion, and in particular the policy ingestion requirement for buildMutex wasn't entirely obvious to me. More details in threads below.

Other than that, I very much appreciate some of the new comments around locking and the assumptions that `regeneratePolicy()` is making :pray: .

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -735,8 +735,18 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
+	policyResult, err := e.regeneratePolicy()
+	stats.policyCalculation.End(err == nil)
+	if err != nil {
+		return false, fmt.Errorf("unable to regenerate policy for '%s': %s", e.StringID(), err)
```
Error formatting should be done with `%w`

christarazi commented on pkg/endpoint/policy.go:
There seem to be some changes in the "endpoint: add policy engine race test" commit that look like they belong in the previous commit.

squeed replied:
:facepalm:

christarazi reviewed (approved):
LGTM pending Joe's comments and a few comments below

joestringer commented on pkg/endpointmanager/manager_test.go:
```diff
@@ -116,45 +118,40 @@ func (epSync *dummyEpSyncher) DeleteK8sCiliumEndpointSync(e *endpoint.Endpoint)
+		name      string
+		setupArgs func() args
+		setupWant func() want
+		//preTestRun func()
```
Why is this commented out? Do we need it at all?

squeed replied:
just an oversight, good catch.

joestringer commented on pkg/endpoint/policy_test.go:
```diff
@@ -38,3 +53,163 @@ func (s *EndpointSuite) TestUpdateVisibilityPolicy(c *check.C) {
+func TestIncrementalUpdatesDuringPolicyGeneration(t *testing.T) {
+	const testfactor = 1000 // bump this to stress-test
+
+	policy.SetPolicyEnabled("always")
```
Incidentally, is this statically defined and could it impact later test runs? :thinking: 

I know in the past we had some weird defer logic for global settings like this to revert the setting back to avoid affecting other tests. I guess if the tests are passing it doesn't directly apply, but it could plausibly trip someone up in future.

squeed replied:
you're right, this should be restored.

joestringer commented on pkg/endpoint/policy_test.go:
```diff
@@ -38,3 +53,163 @@ func (s *EndpointSuite) TestUpdateVisibilityPolicy(c *check.C) {
+	// Create a rule that selects our pod, and allows all peer identities
+	egressDenyRule := &api.Rule{
+		EndpointSelector: api.NewESFromLabels(podSelectLabel),
+		EgressDeny: []api.EgressDenyRule{
```
How does this allow all peer identities if it's an egress deny rule...?

squeed replied:
it's a typo - the word should be "selects", not allows. I had used allow rules but it got coalesced in to a wildcard, neat.

joestringer reviewed (approved):
Just a few remaining minor nits from reviewing the bits I didn't get to before.
